README.rst
..
    This file is part of Invenio.
    Copyright (C) 2015 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

===============
 Invenio-UnAPI
===============

.. image:: https://img.shields.io/travis/inveniosoftware/invenio-unapi.svg
        :target: https://travis-ci.org/inveniosoftware/invenio-unapi

.. image:: https://img.shields.io/coveralls/inveniosoftware/invenio-unapi.svg
        :target: https://coveralls.io/r/inveniosoftware/invenio-unapi

.. image:: https://img.shields.io/github/tag/inveniosoftware/invenio-unapi.svg
        :target: https://github.com/inveniosoftware/invenio-unapi/releases

.. image:: https://img.shields.io/pypi/dm/invenio-unapi.svg
        :target: https://pypi.python.org/pypi/invenio-unapi

.. image:: https://img.shields.io/github/license/inveniosoftware/invenio-unapi.svg
        :target: https://github.com/inveniosoftware/invenio-unapi/blob/master/LICENSE


Invenio module that implements UnAPI service hooks.

*This is an experimental development preview release.*

* Free software: GPLv2 license
* Documentation: https://invenio-unapi.readthedocs.org.


CHANGES.rst
..
    This file is part of Invenio.
    Copyright (C) 2015 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

Changes
=======

Version 0.1.1 (release 2015-10-07):
-----------------------------------

- Upgrades the minimum version of invenio-base to 0.3.0.
- Removes dependencies to invenio.testsuite and replaces them with
  invenio_testing.
- Removes calls to PluginManager consider_setuptools_entrypoints()
  removed in PyTest 2.8.0.
- Adds missing `invenio_base` dependency.

Version 0.1.0 (release 2015-08-20):
-----------------------------------

- Initial public release.


tests/test_unapi.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio
# Copyright (C) 2015 CERN.
#
# Invenio is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

"""Test UnAPI service hooks implementation."""

from __future__ import absolute_import

from flask import url_for

from invenio_testing import InvenioTestCase


class UnAPIMixin(InvenioTestCase):

    """Custom Arxiv configuration."""

    @property
    def config(self):
        """Set format mapping."""
        cfg = super(UnAPIMixin, self).config
        cfg["UNAPI_FORMAT_MAPPING"] = {
            'marcxml': 'xm',
            'dc': 'xd',
        }
        return cfg


class TestUnAPIEndpoint(UnAPIMixin):

    """Test Arxiv query response parsing."""

    def test_index(self):
        response = self.client.get(url_for('unapi.index'))
        assert 'marcxml' in response.data

    def test_redirection(self):
        response = self.client.get(
            url_for('unapi.index', id=1, format='marcxml'))
        assert response.status_code == 302


invenio_unapi/config.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015 CERN.
#
# Invenio is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

"""UnAPI service hooks configuration options."""

from __future__ import unicode_literals

# Define enabled format names fo
UNAPI_FORMAT_MAPPING = {
    'nlm': 'xn',
    'marcxml': 'xm',
    'dc': 'xd',
    'endnote': 'xe',
    'mods': 'xo',
    'xw': 'xw',
}


invenio_unapi/version.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Version information for Invenio-UnAPI.

This file is imported by ``invenio_unapi.__init__``,
and parsed by ``setup.py``.
"""

__version__ = "0.1.2.dev20151007"


invenio_unapi/__init__.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015 CERN.
#
# Invenio is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

"""Invenio module that implements UnAPI service hooks."""

from .version import __version__

__all__ = ('__version__', )


invenio_unapi/views.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2015 CERN.
#
# Invenio is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

"""UnAPI service hooks implementation."""

from flask import (Blueprint, make_response, redirect, render_template,
                   request, url_for)

from invenio_base.globals import cfg
from invenio_formatter.registry import output_formats

from six import iteritems

blueprint = Blueprint('unapi', __name__, url_prefix="/unapi",
                      template_folder='templates', static_folder='static')


@blueprint.route('/', methods=['GET', 'POST'])
def index():
    """Handle /unapi set of pages."""
    # Prepare mappings from configuration.
    mapping = cfg['UNAPI_FORMAT_MAPPING']
    names = {mapping[k]: k for k in mapping}
    # Clean request arguments.
    identifier = request.values.get('id', type=int)
    format_ = request.values.get('format')
    format_ = mapping.get(format_, format_)
    # Redirect to correct record format.
    if identifier and format_:
        return redirect(url_for('record.metadata', recid=identifier,
                                of=format_))

    formats = [{
        'name': names[of],
        'type': values['content_type'],
        'docs': values.get('url'),
    } for of, values in iteritems(output_formats) if of in names]

    response = make_response(render_template(
        'unapi/index.xml', identifier=identifier, formats=formats))
    response.headers['Content-Type'] = 'application/xml'
    return response
